sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+of fundamental data elements. The fundamental data element type **__Ty** must be
+vectorizable, which can be any of the C++ character/integer/real floating-point types,
+or IEEE-754 half-precision floating-point type. The length of the vector
+is a constant expression, as determined by the template parameter **__N** in a
+given specialization.
+
+Each simd class object is mapped to a consecutive block of general register
+files (GRF).
+
+```cpp
+namespace cm {
+namespace gen {
+template <typename __Ty, int __N> class simd {
+public:
+  using value_type = simd<__Ty, __N, 0>;
+  using element_type = __Ty;
+  using vector_type = __vector_type_t<__Ty, __N>;
+  using type = simd<__Ty, __N>;
+  static constexpr int length = __N;
+
+  // Constructors.
+  constexpr simd();
+  constexpr simd(const simd &other);
+  constexpr simd(simd &&other);
+  constexpr simd(const vector_type &__Val);
+  constexpr simd(std::initializer_list<__Ty> __Ilist) noexcept;
+  constexpr simd(__Ty __Val, __Ty __Step = __Ty()) noexcept;
+
+  // Assignment operators.
+  constexpr type &operator=(const type &) &;
+  constexpr type &operator=(type &&) &;
+
+  // Subscript operators.
+  __Ty operator[](int __i);
+  __Ty operator[](int __i) const;
+
+  // Unary operators.
+  type &operator++();
+  type operator++(int);
+  type &operator--();
+  type operator--(int);
+
+  // Binary operators.
+  auto operator +(const type &__RHS) const;
+  auto operator -(const type &__RHS) const;
+  auto operator *(const type &__RHS) const;
+  auto operator /(const type &__RHS) const;
+  type operator &(const type &__RHS) const;
+  type operator |(const type &__RHS) const;
+  type operator ^(const type &__RHS) const;
+
+  // Compound assignment.
+  type &operator +=(const type &__RHS);
+  type &operator -=(const type &__RHS);
+  type &operator *=(const type &__RHS);
+  type &operator /=(const type &__RHS);
+  type &operator &=(const type &__RHS);
+  type &operator |=(const type &__RHS);
+  type &operator ^=(const type &__RHS);
+
+  // Compare operators.
+  simd<uint16_t, __N> operator >(const type &__RHS) const;
+  simd<uint16_t, __N> operator >=(const type &__RHS) const;
+  simd<uint16_t, __N> operator <(const type &__RHS) const;
+  simd<uint16_t, __N> operator <=(const type &__RHS) const;
+  simd<uint16_t, __N> operator ==(const type &__RHS) const;
+  simd<uint16_t, __N> operator !=(const type &__RHS) const;
+
+  // Select operations.
+  template <int __Size, int __Stride>
+  simd<__Ty, __Size> select(uint16_t __Offset = 0) &&;
+  template <int __Size, int __Stride>
+  simd_view<type, region1d_t<__Ty, __Size, __Stride>>
+  select(uint16_t __i = 0) &;
+
+  // Replicate operations.
+  template <int Rep> simd<__Ty, Rep * __N> replicate();
+  template <int Rep, int W> simd<__Ty, Rep * W> replicate(uint16_t __i);
+  template <int Rep, int VS, int W>
+  simd<__Ty, Rep * W> replicate(uint16_t __i);
+  template <int Rep, int VS, int W, int HS>
```

Comment:
Template function with that many arguments are really ugly. Are you sure there isn't a better interface? And that users won't constantly forget what the correct order of the argument is? Some kind of templated builder pattern would be a bit more verbose but much more readable.

Reply:
Can this new builder be added atop later? Can use this comment as a TODO for now.

Reply:
Adding the other interface on-top only works if we want both interfaces. I personally don't. I wouldn't want to have to review code using this interface.

Reply:
OK, can you suggest a change then to converge faster?

Reply:
I suggest that replicate return an (implementation defined) replicate expression. And that allows to add the extra info (width, vstride, hstride). The replicate expression would have those 3 methods (each taking one template-arg) and the implicit conversion to the simd class. The simd class would only have the 1 replicate method taking the single `REP` template-arg and the `i` arg.
The user would use  (any of the extra method calls are optional)
```c++
s.replicate<R>(i).width<W>().vstride<VS>().hstride<HS>()
```

Reply:
I see. This maybe too verbose for some people. I think we should have both - the 4-template arg variant and this one. This one can be added too later.

Reply:
Why do we need a complicated `replicate` function with that many arguments at all? What's the advantage of this complicated function over just asking the user to write it in two steps>
```c++
s.select<W, S1>(i).replicate<R, S2>()
```
If reduced verbosity means it is harder to read it is counter-productive. We should optimize APIs for the reader of code.

Reply:
Replicate is designed to enable programmer to specify arbitrary ISA region supported by Gen HW. Please refer to BSpec description: https://gfxspecs.intel.com/Predator/Home/Index/3010.  The suggested way of writing it in two steps is not equivalent.

Reply:
In addition to @kychendev 's notes above

> Why do we need a complicated replicate function with that many arguments at all?

As I said, `replicate<2,2,2,2>` is much more compact and existing CM users find it quite handy. Plus this not the only place where region spec is used.

> If reduced verbosity means it is harder to read it is counter-productive.

this is quite subjective matter. Regions is a core widely used concept of the Gen SIMD programming. People will quickly get used to it, just be cause it will be used everywhere.

Reply:
This comes back to intention; is this extension to enable existing CM coders to put kernel source in the same TU as their DPC++ host code, or is this to expose Gen HW to users in general?   

If it is the latter, I think "people will get used to it" is not a great argument. I'd recommend that we expose something more verbose/understandable as @rolandschulz suggested and then provide a convenience wrapper to allow the 4-parameter lists that CM users are familiar with.

Reply:
I added a TODO